pkg/internal/stream/stream.go

```diff
@@ -0,0 +1,420 @@
+	if !o.isStopped {
+		o.isStopped = true
+		o.onError(err)
+	}
+}
+
+func (o internalSubscriber[T]) Complete() {
+	if !o.isStopped {
+		o.isStopped = true
+		o.onComplete()
+	}
+}
+
+func safeSubscriber[T any](subcriber types.StreamSubscriber[T]) types.StreamSubscriber[T] {
+	return &internalSubscriber[T]{
+		onNext:     subcriber.Next,
+		onError:    subcriber.Error,
+		onComplete: subcriber.Complete,
+	}
+}
+
+func NewMultiSubscriber[T any](subscribers []types.StreamSubscriber[T]) types.StreamSubscriber[T] {
+	return internalSubscriber[T]{
+		onNext: func(t T) {
+			for _, s := range subscribers {
+				s.Next(t)
+			}
+		},
+		onError: func(err error) {
+			for _, s := range subscribers {
+				s.Error(err)
+			}
+		},
+		onComplete: func() {
+			for _, s := range subscribers {
+				s.Complete()
+			}
+		},
+	}
+}
+
+func operatorSubscriber[T any, U any](dest types.StreamSubscriber[U], onNext func(T)) types.StreamSubscriber[T] {
+	return internalSubscriber[T]{
+		onNext: func(t T) {
+			onNext(t)
+		},
+		onError: func(err error) {
+			dest.Error(err)
+		},
+		onComplete: func() {
+			dest.Complete()
+		},
+	}
+}
+
+type Operator[T any, U any] func(source types.Stream[T], subscriber types.StreamSubscriber[U])
+
+type stream[T any] struct {
+	subscribe func(types.StreamSubscriber[T])
+}
+
+func (s stream[T]) Subscribe(subscriber types.StreamSubscriber[T]) {
+	_, ok := subscriber.(*internalSubscriber[T])
+	if !ok {
+		subscriber = safeSubscriber(subscriber)
+	}
+	s.subscribe(subscriber)
+}
+
+type derivedStream[S any, T any] struct {
+	source   types.Stream[S]
+	operator Operator[S, T]
+}
+
+func (s derivedStream[S, T]) subscribe(subscriber types.StreamSubscriber[T]) {
+	s.operator(s.source, subscriber)
+}
+
+func NewStream[T any](subscribe func(types.StreamSubscriber[T])) types.Stream[T] {
+	return stream[T]{subscribe: subscribe}
+}
+
+func lift[T any, U any](source types.Stream[T], operator Operator[T, U]) types.Stream[U] {
+	return NewStream((derivedStream[T, U]{source: source, operator: operator}).subscribe)
+}
+
+// BufferDebounce groups values after each emission from the source channel for the specified duration
+func BufferDebounce[T any](ctx context.Context, source types.Stream[T], debounceTime time.Duration, clk clock.Clock) types.Stream[[]T] {
+	return lift(source, func(source types.Stream[T], subscriber types.StreamSubscriber[[]T]) {
+		var timerCancel context.CancelFunc
+		var currentValues []T
+		var lock sync.Mutex
+		emit := func() {
+			lock.Lock()
+			if timerCancel != nil {
+				timerCancel()
+			}
+			timerCancel = nil
+			prevValues := currentValues
+			currentValues = nil
+			lock.Unlock()
+			if len(prevValues) != 0 {
+				subscriber.Next(prevValues)
+			}
+		}
+
+		debounceSubscriber := internalSubscriber[T]{
+			onNext: func(next T) {
+				var timerCtx context.Context
+				lock.Lock()
+				currentValues = append(currentValues, next)
+				timerCtx, timerCancel = context.WithCancel(ctx)
```

Comment:
shouldn't you be cancelling any previous to make this a proper debounce? I think you're going to end up with a bunch of timers going off and only one of them being cancelled because only the latest one will be held by `timerCancel`; it'll be safe, but not be doing the kind of debounce batching I think is intended.

```suggestion
				if timerCancel != nil {
					timerCancel()
				}
				timerCtx, timerCancel = context.WithCancel(ctx)
```

Reply:
oh dear you are right.

Reply:
actually I think we need to avoid kicking off a timer at all any time one is running... I think this isn't supposed to be a true debounce per se... it's more of a "if they arrive within a given window together, group them" and actually, we should probably rename this.

Reply:
Yeah, I wondered about that, I think maybe "time batcher" is the kind of thing we want.